crates/sui-move-build/src/linters/redundant_custom_call.rs

```diff
@@ -0,0 +1,267 @@
+        _state: &mut State,
+        _e: &Exp,
+    ) -> Option<Vec<Value>> {
+        None
+    }
+
+    fn call_custom(
+        &self,
+        context: &mut ExecutionContext,
+        _state: &mut State,
+        _loc: &Loc,
+        return_ty: &Type,
+        f: &ModuleCall,
+        args: Vec<Value>,
+    ) -> Option<Vec<Value>> {
+        if let Some((_, _, fname)) = PRIVATE_OBJ_FUNCTIONS
+            .iter()
+            .find(|(addr, module, fun)| f.is(addr, module, fun))
+        {
+            if let Value::LocalObjWithStore(obj_addr_loc) = args[0] {
+                let msg = format!(
+                    "Potential unintended implementation of a custom {} function.",
+                    fname
+                );
+                let (op, action) = if *fname == TRANSFER_FUN {
+                    ("transfer", "transferred")
+                } else if *fname == SHARE_FUN {
+                    ("share", "shared")
+                } else {
+                    ("freeze", "frozen")
+                };
+                let uid_msg = format!("Instances of a type with a store ability can be {} using the \
+                                       public_{} function which often negates the intent of enforcing \
+                                       custom {} policy (implemented through calling the private {} \
+                                       function variant in the module defining this type)", action, fname, op, fname);
+                let mut d = diag!(
+                    REDUNDANT_CUSTOM_DIAG,
+                    (self.fn_name_loc, msg),
+                    (f.name.loc(), uid_msg)
+                );
+                if obj_addr_loc != INVALID_LOC {
+                    let loc_msg = format!("An instance of a module-private type with a store ability to be {} coming from here", action);
+                    d.add_secondary_label((obj_addr_loc, loc_msg));
+                }
+                context.add_diag(d)
+            }
+        }
+        Some(match &return_ty.value {
+            Type_::Unit => vec![],
+            Type_::Single(_) => vec![Value::Other],
+            Type_::Multiple(types) => vec![Value::Other; types.len()],
+        })
+    }
+
+    fn command_custom(&self, _: &mut ExecutionContext, _: &mut State, _: &Command) -> bool {
+        false
+    }
+
+    fn lvalue_custom(
+        &self,
+        _context: &mut ExecutionContext,
+        _state: &mut State,
+        _l: &LValue,
+        _value: &Value,
+    ) -> bool {
+        false
+    }
+}
+
+fn is_local_obj_with_store(st: &SingleType, context: &CFGContext) -> bool {
+    let sp!(_, st_) = st;
+    let bt = match st_ {
+        SingleType_::Base(v) => v,
+        // transfer/share/freeze take objects by value so even if by-reference object has store and
+        // is module-local, it could not end up being an argument to one of these functions
+        SingleType_::Ref(_, _) => return false,
+    };
+    let sp!(_, bt_) = bt;
```

Comment:
Nit: same here for the deconstruction on `bt` -- this deconstruction can just be lifted to the binding of `bt` I believe?